src/github.com/getlantern/lantern-android/client/conn.go

```diff
@@ -0,0 +1,63 @@
+package client
+
+import (
+	"errors"
+	"net"
+)
+
+var (
+	// ErrCouldNotCreateListener is returned when NewListener fails.
+	ErrCouldNotCreateListener = errors.New(`Could not create new listener.`)
+	// ErrClosed is returned when the server is closed.
+	ErrClosed = errors.New(`Server was manually closed.`)
+)
+
+// Listener is a wrapper around net.TCPListener that attempts to provide a
+// Stop() function.
+type Listener struct {
+	*net.TCPListener
+	closed chan bool
+}
+
+// NewListener creates a wrapper around TCPListener
+func NewListener(addr string) (wrap *Listener, err error) {
+	var li net.Listener
+	var tli *net.TCPListener
+
+	var ok bool
+
+	if li, err = net.Listen("tcp", addr); err != nil {
+		return nil, err
+	}
+
+	if tli, ok = li.(*net.TCPListener); !ok {
+		return nil, ErrCouldNotCreateListener
+	}
+
+	wrap = &Listener{
+		TCPListener: tli,
+		closed:      make(chan bool),
+	}
+
+	return wrap, nil
+}
+
+// Accept returns the next connection to the listener.
+func (li *Listener) Accept() (net.Conn, error) {
+	for {
+		select {
+		case <-li.closed:
+			return nil, ErrClosed
+		default:
+			return li.TCPListener.Accept()
+		}
```

Comment:
I don't think this will work quite as intended.  In particular, TCPListener.Accept() will block until a new connection is made.  If no new connection is made, the loop will never get a chance to check li.closed.

Is there a reason you don't just use a regular listener and call its Close() method when you want to stop listening?